samples/drivers/i2s/echo/src/main.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include "codec.h"
+#include <sys/printk.h>
+#include <drivers/i2s.h>
+#include <drivers/gpio.h>
+#include <string.h>
+
+#if DT_NODE_EXISTS(DT_NODELABEL(i2s_rxtx))
+#define I2S_RX_NODE  DT_NODELABEL(i2s_rxtx)
+#define I2S_TX_NODE  I2S_RX_NODE
+#else
+#define I2S_RX_NODE  DT_NODELABEL(i2s_rx)
+#define I2S_TX_NODE  DT_NODELABEL(i2s_tx)
+#endif
+
+#define SAMPLE_FREQUENCY    44100
+#define SAMPLE_BIT_WIDTH    16
+#define BYTES_PER_SAMPLE    sizeof(int16_t)
+#define NUMBER_OF_CHANNELS  2
+/* Such block length provides an echo with the delay of 100 ms. */
+#define SAMPLES_PER_BLOCK   ((SAMPLE_FREQUENCY / 10) * NUMBER_OF_CHANNELS)
+#define INITIAL_BLOCKS      3
+#define TIMEOUT             1000
+
+#define SW0_NODE        DT_ALIAS(sw0)
+#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
+#define SW0_GPIO_CTLR   DT_GPIO_CTLR(SW0_NODE, gpios)
+#define SW0_GPIO_PIN    DT_GPIO_PIN(SW0_NODE, gpios)
+#define SW0_GPIO_FLAGS  DT_GPIO_FLAGS(SW0_NODE, gpios)
+#define SW0_LABEL       DT_PROP(SW0_NODE, label)
+#endif
+
+#define SW1_NODE        DT_ALIAS(sw1)
+#if DT_NODE_HAS_STATUS(SW1_NODE, okay)
+#define SW1_GPIO_CTLR   DT_GPIO_CTLR(SW1_NODE, gpios)
+#define SW1_GPIO_PIN    DT_GPIO_PIN(SW1_NODE, gpios)
+#define SW1_GPIO_FLAGS  DT_GPIO_FLAGS(SW1_NODE, gpios)
+#define SW1_LABEL       DT_PROP(SW1_NODE, label)
+#endif
+
+#define BLOCK_SIZE  (BYTES_PER_SAMPLE * SAMPLES_PER_BLOCK)
+#define BLOCK_COUNT (INITIAL_BLOCKS + 2)
+static K_MEM_SLAB_DEFINE(mem_slab, BLOCK_SIZE, BLOCK_COUNT, 4);
+
+static int16_t echo_block[SAMPLES_PER_BLOCK];
+static volatile bool echo_enabled = true;
+
+#if DT_NODE_HAS_STATUS(SW1_NODE, okay)
+static const struct device *sw1_dev = DEVICE_DT_GET(SW1_GPIO_CTLR);
+#endif
+
+#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
+static void sw0_handler(const struct device *dev, struct gpio_callback *cb,
+			uint32_t pins)
+{
+	bool enable = !echo_enabled;
+
+	echo_enabled = enable;
+	printk("Echo %sabled\n", (enable ? "en" : "dis"));
+}
+#endif
+
+static bool sw1_pressed(void)
+{
+#if DT_NODE_HAS_STATUS(SW1_NODE, okay)
+	return gpio_pin_get(sw1_dev, SW1_GPIO_PIN);
+#else
+	return false;
+#endif
+}
+
+static bool init_buttons(void)
+{
+	int ret;
+
+#if DT_NODE_HAS_STATUS(SW0_NODE, okay)
+	const struct device *sw0_dev = DEVICE_DT_GET(SW0_GPIO_CTLR);
+	static struct gpio_callback sw0_cb_data;
+
+	if (!device_is_ready(sw0_dev)) {
+		printk("%s is not ready\n", sw0_dev->name);
+		return false;
+	}
+
+	ret = gpio_pin_configure(sw0_dev, SW0_GPIO_PIN,
```

Comment:
This would then become a call to gpio_pin_configure_dt(), which saves boilerplate.